toolchain/base/value_store.h

```diff
@@ -189,17 +146,17 @@ class ValueStore
   auto Reserve(size_t size) -> void { values_.reserve(size); }
 
   // These are to support printable structures, and are not guaranteed.
-  auto Print(llvm::raw_ostream& out) const -> void {
-    Print(out, std::nullopt, 0, 0);
-  }
-  auto Print(llvm::raw_ostream& out, std::optional<llvm::StringRef> label,
-             int first_line_indent, int later_indent,
-             // This decays so that `const llvm::APInt` printing catches the
-             // specialization.
-             PrintFn print = Internal::DefaultPrinter()) const -> void {
-    PrintValueRange(out, llvm::iterator_range(values_), label,
-                    first_line_indent, later_indent, /*trailing_newline=*/true,
-                    print);
+  auto OutputYaml() const -> Yaml::OutputMapping {
+    return Yaml::OutputMapping([&](llvm::yaml::IO& io) {
+      for (auto i : llvm::seq(values_.size())) {
+        auto id = IdT(i);
+        Yaml::OutputMapping::Map(
+            io, PrintToString(id),
+            Yaml::OutputScalar([&](llvm::raw_ostream& out) {
+              Internal::PrintValue<ValueT>(out, Get(id));
+            }));
+      }
+    });
```

Comment:
Taking inspiration from the command-line arguments library, I wonder if instead of passing an `IO&` in here, we could pass in an object describing the mapping, so we could instead write this as:

```suggestion
    return Yaml::OutputMapping([&](auto& map) {
      for (auto i : llvm::seq(values_.size())) {
        auto id = IdT(i);
        map.Add(
            PrintToString(id),
            Yaml::OutputScalar([&](llvm::raw_ostream& out) {
              Internal::PrintValue<ValueT>(out, Get(id));
            }));
      }
    });
```

Reply:
Done (though I'm hesitant about `auto` there)